### ISSUE TITLE
SphericalCircle on all-sky plots wraps badly

### ISSUE THREAD

**Issue by keflavich** [astropy/astropy#10780]
SphericalCircle doesn't behave nicely when the circle hits the edge of the map

```
import pylab as pl
from astropy.wcs import WCS
from astropy.io import fits
from astropy.utils.data import get_pkg_data_filename
from astropy.visualization.wcsaxes.frame import EllipticalFrame
from astropy.visualization import simple_norm
from astropy.visualization.wcsaxes import SphericalCircle
from astropy import coordinates, units as u

filename = get_pkg_data_filename('allsky/allsky_rosat.fits')
hdu = fits.open(filename)[0]
wcs = WCS(hdu.header)
data = hdu.data

target_galactic = coordinates.SkyCoord(180.23358104*u.deg, -16.15940703*u.deg, frame='galactic')

pl.figure(figsize=(20,15))
ax = pl.subplot(projection=wcs, frame_class=EllipticalFrame)
im = ax.imshow(data, origin='lower', cmap='gray_r', interpolation='none',
               norm=simple_norm(data, stretch='asinh', max_percent=99.5, min_percent=1))
circle = SphericalCircle((target_galactic.l, target_galactic.b),
                         radius=5*u.deg,
                         transform=ax.get_transform('galactic'))
ax.add_patch(circle)
```

produces

![image](https://user-images.githubusercontent.com/143715/94462103-01f65400-0189-11eb-93d9-d3fefc6da493.png)

**Comment by pllim** [comment]
What is the "not nice" part? The blue band?

**Comment by keflavich** [comment]
Yes, it's supposed to be a circle.  For example, `target_galactic = coordinates.SkyCoord(130.23358104*u.deg, -56.15940703*u.deg, frame='galactic')` creates:
![image](https://user-images.githubusercontent.com/143715/94465712-518b4e80-018e-11eb-91d1-b46b94570144.png)

**Comment by jacopok** [comment]
Hi! I recently encountered this issue as well and I would be open to trying to open a PR with a workaround if it turns out to be feasible.

At the moment my idea is: once the transformed coordinates of the patch are computed, they will have a discontinuity, which causes lines to be drawn across the plot.
If these discontinuities can be located heuristically somehow, the patch can be split into two sections, which should then be drawn independently, probably by overriding the `draw` method in these cases.

I'll look into it, if anyone has a better approach please do let me know.

**Comment by jacopok** [comment]
The MWE I'm using is slightly different from the one in the original issue, without the FITS file and using the plotting interface from `ligo.skymap` since that's my usecase, but I believe that the issue is exactly the same.

```python
import matplotlib.pyplot as plt
from astropy.visualization.wcsaxes import SphericalCircle
from astropy import coordinates, units as u
import ligo.skymap.plot

target = coordinates.SkyCoord(20*u.deg, 0*u.deg)

plt.figure(figsize=(20,15))
ax = plt.subplot(projection='astro hours mollweide')
circle = SphericalCircle((target.ra, target.dec),
                         radius=30*u.deg,
                         transform=ax.get_transform('world')
                         )
ax.add_patch(circle)
```

![mwe_circles](https://github.com/astropy/astropy/assets/11020774/af06277b-a42a-4209-8d7e-8ace1b6b6bf6)

**Comment by neutrinoceros** [comment]
Hi @jacopok, sorry for the late reply. I think the approach you proposed makes sense, designing the heuristic to detect discontinuities/wraparound is probably the difficult part, but without looking too closely I think it should be feasible. Let me know if you are still interested in working on it and/or if you need a hand !

**Comment by jacopok** [comment]
Hi, yes, I haven't worked on it since but if you can lend a hand I'd be happy to try! 

I'm wondering whether this is an astropy-only issue: the code for the [`SphericalCircle`](https://github.com/astropy/astropy/blob/71061de8a5a84b8b31faf7c55d30ff65ebc77e8e/astropy/visualization/wcsaxes/patches.py#L60) does not seem to do anything relevant to this, and the issue comes down to the [`draw` method of the `Patch` object from matplotlib](https://github.com/matplotlib/matplotlib/blob/ff0497cc042332313afd3d2e9ba8f949c3857896/lib/matplotlib/patches.py#L580).

Maybe the generic way to think about it is that when the `Transform` object does wrapping, it could raise a flag, or somehow have a method for detecting where the wrapping happened...  
I'm thinking of something like: the `Transform` needs to wrap the longitudes `[4, 5, 6, 7, 8]`, and it returns `[4, 5, 6, 7-2*pi, 8-2*pi]`. We could have another method, for example `wrap_amount`, that returns something like `[0, 0, 0, 1, 1]` for the previous array. Then in the `draw` method one would simply have to independently plot each separate set of points with a different `wrap_amount`.

I was looking for where that could be implemented, but it confuses me a bit that in the previous examples the transform is from matplotlib, a [`CompositeGenericTransform`](https://github.com/matplotlib/matplotlib/blob/ff0497cc042332313afd3d2e9ba8f949c3857896/lib/matplotlib/transforms.py#L2348) - shouldn't it be something from the wcs module in astropy?

**Comment by neutrinoceros** [comment]
This all seem reasonable.

> I was looking for where that could be implemented, but it confuses me a bit that in the previous examples the transform is from matplotlib, a [CompositeGenericTransform](https://github.com/matplotlib/matplotlib/blob/ff0497cc042332313afd3d2e9ba8f949c3857896/lib/matplotlib/transforms.py#L2348) - shouldn't it be something from the wcs module in astropy?

can you specify what `Transform` object you're referring to ?
While I'm here, can you tell me how to install `ligo` so I can reproduce your simplified example ?

**Comment by jacopok** [comment]
I mean the Transform object you can get with `circle.get_transform()` in the previous examples, I understood it to be responsible for the coordinate transformation that causes the issue.

The package is actually `ligo.skymap`, you can find a reference for its installation here: 
https://lscsoft.docs.ligo.org/ligo.skymap/quickstart/install.html
